File: playground/cpp/days/day03/tests/ticket_dispenser_test.cpp
#include "day03/ticket_dispenser.h"
#include "run_together.h"
#include <gtest/gtest.h>

#include <algorithm>
#include <atomic>
#include <vector>

TEST(TicketDispenser, FirstTicketIsOne) {
    TicketDispenser d;
    EXPECT_EQ(d.next(), 1);
}

TEST(TicketDispenser, TicketsCountUpByOne) {
    TicketDispenser d;
    EXPECT_EQ(d.next(), 1);
    EXPECT_EQ(d.next(), 2);
    EXPECT_EQ(d.next(), 3);
}

TEST(TicketDispenser, IssuedCountsHandedOutTickets) {
    TicketDispenser d;
    EXPECT_EQ(d.issued(), 0);
    d.next();
    d.next();
    EXPECT_EQ(d.issued(), 2);
}

TEST(TicketDispenser, ConcurrentTicketsAreUniqueAndGapless) {
    constexpr int kThreads = 8;
    constexpr int kPerThread = 20000;
    TicketDispenser d;
    std::vector<std::vector<long long>> got(kThreads);
    runTogether(kThreads, [&](int t) {
        for (int i = 0; i < kPerThread; ++i) got[t].push_back(d.next());
    });

    std::vector<long long> all;
    for (const auto& g : got) all.insert(all.end(), g.begin(), g.end());
    std::sort(all.begin(), all.end());
    for (std::size_t i = 0; i < all.size(); ++i) {
        ASSERT_EQ(all[i], static_cast<long long>(i) + 1)
            << "a ticket was handed out twice, or a number was skipped";
    }
    EXPECT_EQ(d.issued(), kThreads * kPerThread);
}

TEST(TicketDispenser, EachThreadSeesItsTicketsIncrease) {
    constexpr int kThreads = 8;
    constexpr int kPerThread = 20000;
    TicketDispenser d;
    std::vector<std::vector<long long>> got(kThreads);
    runTogether(kThreads, [&](int t) {
        for (int i = 0; i < kPerThread; ++i) got[t].push_back(d.next());
    });

    for (const auto& g : got) {
        for (std::size_t i = 1; i < g.size(); ++i) {
            ASSERT_LT(g[i - 1], g[i]) << "a later ticket had a smaller number";
        }
    }
}

TEST(TicketDispenser, IssuedNeverGoesBackwardsWhileTicketsAreTaken) {
    constexpr int kTakers = 4;
    constexpr int kPerThread = 20000;
    TicketDispenser d;
    std::atomic<bool> wentBackwards{false};
    runTogether(kTakers + 1, [&](int t) {
        if (t == kTakers) {  // the watcher
            long long last = 0;
            for (int i = 0; i < 20000; ++i) {
                long long now = d.issued();
                if (now < last) wentBackwards = true;
                last = now;
            }
            return;
        }
        for (int i = 0; i < kPerThread; ++i) d.next();
    });

    EXPECT_FALSE(wentBackwards);
    EXPECT_EQ(d.issued(), kTakers * kPerThread);
}

File: playground/java/days/day03/src/test/java/day03/TicketDispenserTest.java
package day03;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;

import java.util.Arrays;
import java.util.concurrent.atomic.AtomicBoolean;

import static day03.RunTogether.runTogether;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

@Timeout(value = 60, threadMode = Timeout.ThreadMode.SEPARATE_THREAD)
class TicketDispenserTest {

    @Test
    void firstTicketIsOne() {
        TicketDispenser d = new TicketDispenser();
        assertEquals(1, d.next());
    }

    @Test
    void ticketsCountUpByOne() {
        TicketDispenser d = new TicketDispenser();
        assertEquals(1, d.next());
        assertEquals(2, d.next());
        assertEquals(3, d.next());
    }

    @Test
    void issuedCountsHandedOutTickets() {
        TicketDispenser d = new TicketDispenser();
        assertEquals(0, d.issued());
        d.next();
        d.next();
        assertEquals(2, d.issued());
    }

    @Test
    void concurrentTicketsAreUniqueAndGapless() throws Exception {
        int threads = 8;
        int perThread = 20_000;
        TicketDispenser d = new TicketDispenser();
        long[][] got = new long[threads][perThread];
        runTogether(threads, t -> {
            for (int i = 0; i < perThread; i++) got[t][i] = d.next();
        });

        long[] all = Arrays.stream(got).flatMapToLong(Arrays::stream).sorted().toArray();
        for (int i = 0; i < all.length; i++) {
            assertEquals(i + 1, all[i], "a ticket was handed out twice, or a number was skipped");
        }
        assertEquals(threads * perThread, d.issued());
    }

    @Test
    void eachThreadSeesItsTicketsIncrease() throws Exception {
        int threads = 8;
        int perThread = 20_000;
        TicketDispenser d = new TicketDispenser();
        long[][] got = new long[threads][perThread];
        runTogether(threads, t -> {
            for (int i = 0; i < perThread; i++) got[t][i] = d.next();
        });

        for (long[] g : got) {
            for (int i = 1; i < g.length; i++) {
                assertTrue(g[i - 1] < g[i], "a later ticket had a smaller number");
            }
        }
    }

    @Test
    void issuedNeverGoesBackwardsWhileTicketsAreTaken() throws Exception {
        int takers = 4;
        int perThread = 20_000;
        TicketDispenser d = new TicketDispenser();
        AtomicBoolean wentBackwards = new AtomicBoolean(false);
        runTogether(takers + 1, t -> {
            if (t == takers) { // the watcher
                long last = 0;
                for (int i = 0; i < 20_000; i++) {
                    long now = d.issued();
                    if (now < last) wentBackwards.set(true);
                    last = now;
                }
                return;
            }
            for (int i = 0; i < perThread; i++) d.next();
        });

        assertFalse(wentBackwards.get(), "issued() went backwards while tickets were taken");
        assertEquals(takers * perThread, d.issued());
    }
}
